src/com/google/ProfilingReport.java

```diff
@@ -0,0 +1,249 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+package com.google;
+
+import com.google.ProductionInstrumentationReporter.InstrumentationType;
+import com.google.auto.value.AutoValue;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Predicate;
+
+
+/**
+ * A class that maintains all information about the production instrumentation results which will be
+ * converted to a JSON.
+ */
+public class ProfilingReport {
+
+  int totalReportsParsed;
+  float percentOfFunctionsExecuted;
+  float percentOfBranchesExecuted;
+
+  List<FileProfilingResult> result;
+
+
+  private ProfilingReport(int totalReportsParsed, List<FileProfilingResult> result) {
+    this.totalReportsParsed = totalReportsParsed;
+    this.result = result;
+
+    calculateProfilingReportPercents();
+  }
+
+  private static float roundFloat2Decimals(float num) {
+    return (Math.round(num * 10000) / (float) 100);
+  }
+
+  /**
+   * This function takes the instrumentationMapping and reports sent by the instrumented production
+   * code and creates a list of ProfilingResultsByFile with the aggregated information. It will
+   * average the ProfilingData provided by allInstrumentationReports and also calculate a percentage
+   * of how often a parameter is encounter.
+   *
+   * @param instrumentationMapping    The instrumentationMapping generated by the Production
+   *                                  Instrumentation pass
+   * @param allInstrumentationReports A list off all reports sent by the instrumented production
+   *                                  code
+   * @return A list of ProfilingResultsByFile where each element is the profiling result for a
+   * different source file.
+   */
+  public static ProfilingReport createProfilingReport(
+      InstrumentationMapping instrumentationMapping,
+      List<Map<String, ProfilingData>> allInstrumentationReports) {
+
+    List<FileProfilingResult> result = new ArrayList<>();
+
+    // Iterate over all fileNames since that is what we are grouping by.
+    for (String fileName : instrumentationMapping.fileNames) {
+
+      // Get all instrumentation parameters where the source file is fileName.
+      List<String> instrumentationPointsPerFile = instrumentationMapping.getAllMatchingValues((s) ->
+          fileName.equals(instrumentationMapping.getFileName(s))
+      );
+
+      Map<String, List<ProfilingResult>> profilingDataPerFunction = new HashMap<>();
+
+      for (String param : instrumentationPointsPerFile) {
```

Comment:
I think breaking up this method would help readability. In particular, I find the nested for loop and if block below too nested (four levels of nesting gets confusing). This four-loop might be a good place to start.

See also https://dzone.com/articles/code-smells-deeply-nested-code

Reply:
Broke down the for-loops into three different functions